Fix pruning logic in ConversationSummaryBufferMemory for persistent m…

joybro opened the pull request:
…essage histories

This commit addresses the issue where the pruning logic in ConversationSummaryBufferMemory did not work as expected with persistent message histories like FileChatMessageHistory.

Changes made:
1. Added set_messages() to BaseChatMessageHistory to store updated history after pruning instead of relying on List.pop(), which does not work for persistent memory histories.

Now, the ConversationSummaryBufferMemory pruning mechanism correctly handles persistent message histories and ensures that the message history stays within the max_token_limit.

Fixes #3072

hwchase17 reviewed (commented):
probably more direct to implement a `pop` method on each history object

any reason to go this way vs considering that?

joybro committed:
Fix pruning logic in ConversationSummaryBufferMemory for persistent message histories

This commit addresses the issue where the pruning logic in
ConversationSummaryBufferMemory did not work as expected with
persistent message histories like FileChatMessageHistory.

Changes made:
1. Added pop() to BaseChatMessageHistory to remove and return the message
   at the given index. This enables ConversationSummaryBufferMemory to
   correctly update the message history after pruning, regardless of
   whether the history is persistent or not.

Now, the ConversationSummaryBufferMemory pruning mechanism correctly
handles persistent message histories and ensures that the message
history stays within the max_token_limit.

Fixes hwchase17#3072

joybro commented:
@hwchase17, you're right. Initially, I implemented the set_messages() method to address another issue that I haven't reported yet, which is that ConversationSummaryBufferMemory does not persist its summary across sessions.

However, I agree that it might be more straightforward to fix this issue using the pop() method first, and then discuss the other issue and its solution separately. I'll update my changes and submit a new PR. Thanks for having a look!

dev2049 commented on langchain/memory/chat_message_histories/file.py:
```diff
@@ -28,6 +28,13 @@ def __init__(self, file_path: str):
+                json.loads(self.file_path.read_text())
+            except json.JSONDecodeError:
+                # If it fails, initialize the file with an empty list
+                self.file_path.write_text(json.dumps([]))
```
so this is overwriting the original file? would it be better to create a new file?

uwej711 commented:
Could this be done differently? We would like to store the complete interaction in the database for humans to read. Also, since here pop is not implemented for all memory types, that would render this summarizer useless with that memory.

uwej711 commented:
Instead of actually pruning the messages we could keep the indices of the messages in the buffer to use.

raffertyuy commented:
Hi @joybro and @hwchase17, encountered this issue today and wondering how it's going, as the pull request was since april.
Also, it looks like pop() is only implemented for `chat_message_histories` → `file.py` and `in_memory.py`?

rapcal commented:
Any recommendations on how to deal with this? Got hit by this issue in an application running in production, using Redis as the memory store... 🤯

leo-gan commented:
@joybro Hi , could you, please, resolve the merging issues and address the last comments (if needed)? After that ping me and I push this PR for the review. Thanks!

efriis commented:
Closing because the PR wouldn't line up with the current directory structure of the library (would need to be in /libs/langchain/langchain instead of /langchain). Feel free to reopen against the current head if it's still relevant!
